### ISSUE TITLE
Crashes in hiveserver2 -> get_operation_status()

### ISSUE THREAD

**Issue by natekupp** [cloudera/impyla#135]
We are often seeing crashes in impyla down in `get_operation_status` (note that the line # corresponds to 0.9.1, I believe this is [hiveserver2.py#865](https://github.com/cloudera/impyla/blob/master/impala/hiveserver2.py#L865) on master. 

The problem seems to be that `service.GetOperationStatus(req)` decides the response is ok, but `resp.operationState` is `None`, causing the dictionary lookup in `_VALUES_TO_NAMES` to fail.

What's interesting is that the query appears to have been successful at this point (so far we haven't seen any crashes where the query was not successful). 

See example traceback below.

    Traceback (most recent call last):
    ...
        cur = self.conn.cursor()
      File "/usr/local/lib/python2.7/dist-packages/impala/dbapi/hiveserver2.py", line 59, in cursor
        cursor.execute('USE %s' % self.default_db)
      File "/usr/local/lib/python2.7/dist-packages/impala/dbapi/hiveserver2.py", line 152, in execute
        self._execute_sync(op)
      File "/usr/local/lib/python2.7/dist-packages/impala/dbapi/hiveserver2.py", line 160, in _execute_sync
        self._wait_to_finish()  # make execute synchronous
      File "/usr/local/lib/python2.7/dist-packages/impala/dbapi/hiveserver2.py", line 179, in _wait_to_finish
        self.service, self._last_operation_handle)
      File "/usr/local/lib/python2.7/dist-packages/impala/_rpc/hiveserver2.py", line 132, in wrapper
        return func(*args, **kwargs)
      File "/usr/local/lib/python2.7/dist-packages/impala/_rpc/hiveserver2.py", line 411, in get_operation_status
        return TOperationState._VALUES_TO_NAMES[resp.operationState]
    KeyError: None

**Comment by laserson** [comment]
This is interesting.  The query state machine is not well documented, so I need to check with the Impala team to see what circumstances could cause a `None` to be returned in the response.  Can you reliably reproduce this?  IIRC, some of this logic has been modified in later versions of impyla.  Can you also try upgrading to the latest version?

**Comment by laserson** [comment]
Perhaps this branch is causing the empty `operationState`:
https://github.com/cloudera/Impala/blob/cdh5-trunk/be/src/service/impala-hs2-server.cc#L925

Is it possible these queries are getting cleaned up by something else before this check happens?

**Comment by laserson** [comment]
Actually, even if the `operationState` is unset, the `TStatus` should've been set to some kind of error state, and `err_if_rpc_not_ok` should catch it.  It would be interesting to see what the associated `TStatus` object looks like.  If you use the latest impyla, you can turn on debug logging, which would dump all of the thrift messages into a log.

**Comment by natekupp** [comment]
It seems like this was due to a bad interaction of impyla and a multiprocessing pool. Essentially, we were sharing the result of calling `impala.dbapi.connect()`  across subprocesses, and then creating a `cur = self.conn.cursor()` within each subprocess.

Here's a stripped down example:

https://gist.github.com/natekupp/33bdc7b7324e68afbe8f

We've moved the instantiation of the impala object to within a run() function called by the multiprocessing pool, which seems to have resolved this crash.
